containers-tests/tests/tree-properties.hs

```diff
@@ -179,12 +211,28 @@ prop_product :: Tree OrdA -> Property
 prop_product t = product t === product (toList t)
 
 #if MIN_VERSION_base(4,18,0)
+-- We use Magma with foldMap1 to test that the structure of the fold follows
+-- that of the tree. This is a desirable property to have.
+prop_foldMap1_structure :: Tree A -> Property
+prop_foldMap1_structure t =
+  Foldable1.foldMap1 Inj t === foldTree (\x ys -> Foldable1.fold1 (Inj x :| ys)) t
+
 prop_toNonEmpty :: Tree A -> Property
 prop_toNonEmpty t = Foldable1.toNonEmpty t === NE.fromList (toList t)
 
 prop_last :: Tree A -> Property
 prop_last t = Foldable1.last t === NE.last (Foldable1.toNonEmpty t)
 
+-- Tests that last only looks at the path going down to the last leaf.
+prop_last_path :: Tree A -> Property
+prop_last_path t = Foldable1.last (replace t) === Foldable1.last t
+  where
+    -- Replace all trees with bottom except for the last one.
+    replace (Node x (t:ts)) = Node x ts'
+      where
+        ts' = Foldable1.foldrMap1 ((:[]) . replace) (const (error "error tree" :)) (t :| ts)
+    replace t@(Node _ []) = t
```

Comment:
I can't read this. Maybe that's just me, but I suspect we can be clearer. The following will be slightly less efficient, but I don't think we have to worry about that too much for this test.

```haskell
replace (Node x ts) = Node x (replaces ts)
replaces :: [Tree a] -> [Tree a]
replaces [] = []
replaces [t] = [replace t]
replaces (t : ts) = error "error tree" : replaces ts
```

Reply:
Sure, done.